openraft/src/core/mod.rs

```diff
@@ -250,56 +249,69 @@ impl<D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>> Ra
     async fn do_main(&mut self) -> Result<(), Fatal> {
         tracing::debug!("raft node is initializing");
 
-        // NOTE: get_initial_state will return None, if Raft node is first startup.
-        let state = if let Some(init_state) = self.storage.get_initial_state().await? {
-            init_state
-        } else {
-            let init_state = InitialState::new(self.id);
-            self.storage.save_hard_state(&init_state.hard_state).await?;
-            init_state
-        };
+        let state = self.storage.get_initial_state().await?;
+
+        // TODO(xp): this is not necessary.
+        self.storage.save_hard_state(&state.hard_state).await?;
 
         self.last_log_id = state.last_log_id;
         self.current_term = state.hard_state.current_term;
         self.voted_for = state.hard_state.voted_for;
-        self.effective_membership = state.last_membership.clone();
+        self.effective_membership = state.last_membership.unwrap_or_else(|| EffectiveMembership::new_initial(self.id));
         self.last_applied = state.last_applied;
 
-        // NOTE: this is repeated here for clarity. It is unsafe to initialize the node's commit
-        // index to any other value. The commit index must be determined by a leader after
+        // NOTE: The commit index must be determined by a leader after
         // successfully committing a new log to the cluster.
         self.committed = None;
 
         // Fetch the most recent snapshot in the system.
         if let Some(snapshot) = self.storage.get_current_snapshot().await? {
             self.snapshot_last_log_id = Some(snapshot.meta.last_log_id);
-            self.report_metrics(Update::Ignore);
+            self.report_metrics(Update::AsIs);
         }
 
-        let has_log = self.last_log_id.is_some();
-        let single = self.effective_membership.membership.all_nodes().len() == 1;
-        let is_voter = self.effective_membership.membership.contains(&self.id);
+        let has_log = if self.last_log_id.is_some() {
+            "has_log"
+        } else {
+            "no_log"
+        };
+
+        let single = if self.effective_membership.membership.all_nodes().len() == 1 {
+            "single"
+        } else {
+            "multi"
+        };
+
+        let is_voter = if self.effective_membership.membership.contains(&self.id) {
+            "voter"
+        } else {
+            "learner"
+        };
 
         self.target_state = match (has_log, single, is_voter) {
             // A restarted raft that already received some logs but was not yet added to a cluster.
             // It should remain in Learner state, not Follower.
-            (true, true, false) => State::Learner,
-            (true, false, false) => State::Learner,
+            ("has_log", "single", "learner") => State::Learner,
```

Comment:
maybe it is better use if instead of match in this case more readable. ```rust if has_log && !is_voter { self.target_state = State::Learner  } ``` . If you agree, i can do that later

Reply:
It was the if-else style before. But the branching is hard to make it short and at the same time easy to see if it exhausts all of the combinations.
And there was a bug with the unexhausted condition: https://github.com/datafuselabs/openraft/commit/39690593a07c6b9ded4b7b8f1aca3191fa7641e4#diff-bffbf2e4f7d33dabb4d70c100a1293c841e23c2a2dc3d2174ea9b1e04d200bdc